serialization/python_pickle.ksy

```diff
@@ -0,0 +1,588 @@
+meta:
+  id: python_pickle
+  title: Python pickle serialization format
+  application: Python
+  file-extension:
+    - pickle
+    - pkl
+  license: CC0-1.0
+  endian: le
+  xref:
+    justsolve: Pickle
+    wikidata: Q7190889
+doc: |
+  Python Pickle format serializes Python objects to a byte stream, as a sequence
+  of operations to run on the Pickle Virtual Machine.
+
+  The format is mostly implementation defined, there is no formal specification.
+  Pickle data types are closely coupled to the Python object model.
+  Python singletons, and most builtin types (e.g. `None`, `int`,`dict`, `list`)
+  are serialised using dedicated Pickle opcodes.
+  Other builtin types, and all classes  (e.g. `set`, `datetime.datetime`) are
+  serialised by encoding the name of a constructor callable.
+  They are deserialised by importing that constructor, and calling it.
+  So, unpickling an arbitrary pickle, using the Python's stdlib pickle module
+  can cause arbitrary code execution.
+
+  Pickle format has evolved with Python, later protocols add opcodes & types.
+  Later Python releases can pickle to or unpickle from any earlier protocol.
+
+  * Protocol 0: ASCII clean, no explicit version, fields are '\n' terminated.
+  * Protocol 1: Binary, no explicit version, first length prefixed types.
+  * Protocol 2: Python 2.3+. Explicit versioning, more length prefixed types.
+    https://www.python.org/dev/peps/pep-0307/
+  * Protocol 3: Python 3.0+. Dedicated opcodes for `bytes` objects.
+  * Protocol 4: Python 3.4+. Opcodes for 64 bit strings, framing, `set`.
+    https://www.python.org/dev/peps/pep-3154/
+doc-ref: https://github.com/python/cpython/blob/v3.7.3/Lib/pickletools.py
+seq:
+  # TODO is there a way to declare PROTO is optional, but only valid at position 0?
+  - id: ops
+    type: op
+    repeat: eos
+  # TODO is there a way to declare a trailing STOP is required?
```

Comment:
As I have said, there is. Somethkng like these, though there may be better ways to express this, I just have no time to play with it now:

```yaml
seq:
  - id: placeholder # to create a substream
    size-eos: true # if I remember right
instances:
  main:
    io: placeholder._io
    pos: 0
    size: placeholder._size -1
  assert_stop:
    pos: placeholder._size -1
    type: assert_stop
types:
  placeholder: {}
  assert_stop:
    seq:
      - contents: ["."]
```

Reply:
Similarly to `PROTO` I now think this is better handled outside this format declaration (and parsers generated from it). Sorry for leaving it hanging so long that you had to write the above.

If you would prefer to enforce this here, then I will try to make it work.